### ISSUE TITLE
Unhelpful error message: expected "NDArray[Unknown]" but received "NDArray[Unknown]"

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#6531]
I encountered the error message (in a rather complex example)

```
error: "assert_type" mismatch: expected "NDArray[Unknown]" but received "NDArray[Unknown]"
```

The issue seems to be related to the `TypeVar` not being bound to `np.generic`. Below is a MWE reproduction of the issue.

```python
from typing import Any, Generic, Protocol, Sequence, TypeVar, assert_type

import numpy as np
from numpy.typing import NDArray


class Unknown(Protocol):
    """Dummy protocol."""


T = TypeVar("T", bound=Unknown)


class Foo(Generic[T]):
    """Dummy class."""

    data: NDArray[T]

    def __init__(self, data: Sequence[T]):
        self.data = np.array(data)

    def where_greater_zero(self) -> NDArray[T]:
        zero = self.data[0] * 0
        return self.data[self.data > zero]


obj: Foo[Any] = Foo([1, 2, 3])
assert_type(obj.where_greater_zero(), NDArray)
```

**Comment by erictraut** [comment]
Thanks for the issue. I've made a couple of changes here. First, I've improved the error-handling when the explicit specialization of a generic type argument fails. The old logic was producing a type that filled in `Unknown` for the type alias but not for the underlying type associated with it. This is what led to the strange behavior with `assert_type`. Second, I've improved `assert_type` so it expands type aliases when reporting a type mismatch. This should help identify the cause of the mismatch.
